### DISCUSSION TITLE
BatchNorm don't work as expected

### DISCUSSION THREAD

**Discussion by wangjiawen2013** [tracel-ai/burn#2650] [Q&A] [answered]
Hi,
Here is the output of pytorch `BatchNorm1d`, it worked well:
![image](https://github.com/user-attachments/assets/52ba0ff6-29d1-499d-bef4-fed01ddb48e7)

and here is the rust burn `BatchNorm`, the output is weired:
![image](https://github.com/user-attachments/assets/cff494d6-7336-43ef-90b0-e0724a0a9b7e)

Then I changed to `AutodiffBackend`, the results is correct !  Why don't limit the `BatchNorm` to only `AutodiffBackend` ? Otherwise we are prone to make mistakes.
![image](https://github.com/user-attachments/assets/3dba26d0-58a2-4014-978e-cb058c9f25aa)

**Comment by laggui** [comment, marked as answer]
The result is not wrong 🙂 

The difference lies in the training vs inference computation for a batchnorm module. If you use `m.eval()` instead for the pytorch module you should get equivalent results.

With burn you have to be explicit when using autodiff. In pytorch it's kind of the opposite, it will track the gradients and keep the autodiff graph by default unless you use `with torch.no_grad()` context.

This is also explained in the [autodiff section](https://burn.dev/burn-book/building-blocks/autodiff.html#difference-with-pytorch).

**Reply by dfsfdfse** [reply]
the BatchNorm parameters yield different results in burn compared to pytorch。
the running state parameters are also included in the calculation.

**Comment by wangjiawen2013** [comment]
Good explanation. Here is a relate description from pytorch:

![image](https://github.com/user-attachments/assets/c239289c-3741-43d2-85c8-5f814e3c9e3b)

Is `dropout` the same as `batchnorm` in burn from this view ?

**Reply by laggui** [reply]
Both dropout and batchnorm have different behavior when training.

Dropout will only be applied when using autodiff. For inference, it doesn't make sense 🙂
